File: Easy/Valid Parentheses/validparentheses.cpp
#include <list>
#include <unordered_map> //hash table DS

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        list<char> stack = {};
        unordered_map<char,char> parentheses = {{'(',')'},{'[',']'},{'{','}'}};
        for (char c: s){
            if (parentheses.count(c) == 1){ //if c is an opening parenthesis
                stack.push_back(c);
            }
            else {
                if (stack.size() > 0 && parentheses[stack.back()] == c){
                    stack.pop_back();
                }
                else return false;
            }
        }
        //if there are leftover opening parentheses in the stack, the input stirng is not valid
        return stack.size() > 0 ? false : true;
    }
};

File: Easy/Valid Parentheses/validparentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []; parentheses = {"(":")", "[":"]", "{":"}"}
        for char in s:
            if char in parentheses: #if character is opening parenthesis
                stack.append(char)
            else: #closing parenthesis
                if stack and parentheses[stack[-1]] == char: #check if stack is empty
                    stack.pop()
                else: #parentheses do not match
                    return False
        #if there are leftover opening parentheses in the stack, the input stirng is not valid
        return False if stack else True
